File: DataStructures/src/main/java/com/zhangdd/sparsearray/SparseArray.java
package com.zhangdd.sparsearray;

/**
 * @author zhangdd on 2021/9/22
 */
public class SparseArray {

    public static void main(String[] args) {
        //1.定义二维数组，并设置二维数组中有的数据
        int chessData[][] = new int[11][11];
        chessData[1][2] = 1;
        chessData[2][3] = 2;
        chessData[4][3] = 1;

        //2.记录有效值的个数
        int effectiveNumber = 0;
        for (int i = 0; i < chessData.length; i++) {
            for (int j = 0; j < chessData[i].length; j++) {
                int data = chessData[i][j];
                System.out.printf("%d\t", data);
                if (data != 0) {
                    effectiveNumber++;
                }
            }
            System.out.println();
        }

        //3.创建稀疏数组
        int sparseArray[][] = new int[effectiveNumber + 1][3];
        sparseArray[0][0] = 11;
        sparseArray[0][1] = 11;
        sparseArray[0][2] = effectiveNumber;

        //4.为稀疏数组赋值
        int record = 0;
        for (int i = 0; i < chessData.length; i++) {
            for (int j = 0; j < chessData[i].length; j++) {
                int data = chessData[i][j];
                if (data != 0) {
                    record++;
                    sparseArray[record][0] = i;
                    sparseArray[record][1] = j;
                    sparseArray[record][2] = chessData[i][j];
                }
            }
        }

        for (int i = 0; i < sparseArray.length; i++) {
            System.out.printf("%d\t,%d\t,%d\t\n", sparseArray[i][0], sparseArray[i][1], sparseArray[i][2]);
        }

        //5.将稀疏数组恢复为原来数组
        int newChess[][] = new int[sparseArray[0][0]][sparseArray[0][1]];
        //6.为原始数组填充数据
        for (int i = 1; i < sparseArray.length; i++) {
            newChess[sparseArray[i][0]][sparseArray[i][1]] = sparseArray[i][2];
        }

        for (int i = 0; i < newChess.length; i++) {
            for (int j = 0; j < newChess[i].length; j++) {
                int data = chessData[i][j];
                System.out.printf("%d\t", data);
            }
            System.out.println();
        }

    }
}

File: python/DataStructures/sparsearry/SparseArray.py
#!/usr/bin/env python3
"""
@author: zhangdd on 2022/1/4
"""
# 定义一个11*11的二维数组
chess_data = [[None for i in range(11)] for j in range(11)]
# 初始化数据
chess_data[1][2] = 1
chess_data[2][3] = 2
chess_data[4][3] = 1

valid_value_count = 0

for row in chess_data:
    for item in row:
        if item is None:
            pass
        else:
            valid_value_count += 1

# 定义稀疏数组
sparse_array = [[None for i in range(3)] for j in range(valid_value_count + 1)]
# 初始化稀疏数组数据
sparse_array[0][0] = len(chess_data)
sparse_array[0][1] = len(chess_data[0])
sparse_array[0][2] = valid_value_count + 1

sparse_index = 1
for i in range(len(chess_data)):
    for j in range(len(chess_data[i])):
        if chess_data[i - 1][j - 1] is None:
            pass
        else:
            sparse_array[sparse_index][0] = i - 1
            sparse_array[sparse_index][1] = j - 1
            sparse_array[sparse_index][2] = chess_data[i - 1][j - 1]
            sparse_index += 1

print(sparse_array)
